source/credentials_provider_imds.c

```diff
@@ -653,17 +756,50 @@ static void s_imds_on_acquire_connection(struct aws_http_connection *connection,
             error_code,
             aws_error_str(error_code));
 
-        s_imds_finalize_get_credentials_query(imds_user_data);
-        return;
+        imds_user_data->query_state = AWS_IMDS_QS_UNRECOVERABLE_ERROR;
     }
 
-    imds_user_data->connection = connection;
-    if (imds_user_data->token_required) {
-        imds_user_data->query_state = AWS_IMDS_QS_TOKEN;
-        s_imds_query_token(imds_user_data);
+    s_query_state_machine[imds_user_data->query_state](imds_user_data);
+    /* there's no universe where we should have moved to COMPLETE, but an error could have occurred. */
+    if (imds_user_data->query_state == AWS_IMDS_QS_UNRECOVERABLE_ERROR) {
+        s_query_state_machine[AWS_IMDS_QS_UNRECOVERABLE_ERROR](imds_user_data);
+        s_aws_credentials_provider_imds_user_data_destroy(imds_user_data);
+    }
+}
+
+static void s_imds_on_stream_complete_fn(struct aws_http_stream *stream, int error_code, void *user_data) {
+    struct aws_credentials_provider_imds_user_data *imds_user_data = user_data;
+
+    aws_http_message_destroy(imds_user_data->request);
+    imds_user_data->request = NULL;
+
+    struct aws_credentials_provider_imds_impl *impl = imds_user_data->imds_provider->impl;
+
+    struct aws_http_connection *connection = impl->function_table->aws_http_stream_get_connection(stream);
+    impl->function_table->aws_http_stream_release(stream);
+    impl->function_table->aws_http_connection_manager_release_connection(impl->connection_manager, connection);
+
+    /* try again, just drop the state from the response to the request state by subtracting one.
+     * Don't run the state machine in this callback in this case, let the acquire connection callback handle it. */
+    if (error_code == AWS_ERROR_HTTP_CONNECTION_CLOSED) {
+        AWS_LOGF_DEBUG(
```

Comment:
Glad to see this here (was about to complain about its absence) but I feel like this needs a disclaimer along the lines of "temporary/TODO: add retry strategy hook when that system gets implemented"

Reply:
Actually, is there a not-terrible temporary hack we can do here to make sure we always terminate in a reasonable amount of time even in the face of something pathological (repeated connection closes)?

Reply:
I'll put in a max retry count.